src/errno.cr

```diff
@@ -18,20 +18,8 @@ end
 # dealing with C libraries.
 enum Errno
   NONE = 0
-
-  {% for value in %w(E2BIG EPERM ENOENT ESRCH EINTR EIO ENXIO ENOEXEC EBADF ECHILD EDEADLK ENOMEM
-                    EACCES EFAULT ENOTBLK EBUSY EEXIST EXDEV ENODEV ENOTDIR EISDIR EINVAL ENFILE
-                    EMFILE ENOTTY ETXTBSY EFBIG ENOSPC ESPIPE EROFS EMLINK EPIPE EDOM ERANGE EAGAIN
-                    EWOULDBLOCK EINPROGRESS EALREADY ENOTSOCK EDESTADDRREQ EMSGSIZE EPROTOTYPE ENOPROTOOPT
-                    EPROTONOSUPPORT ESOCKTNOSUPPORT EPFNOSUPPORT EAFNOSUPPORT EADDRINUSE EADDRNOTAVAIL
-                    ENETDOWN ENETUNREACH ENETRESET ECONNABORTED ECONNRESET ENOBUFS EISCONN ENOTCONN
-                    ESHUTDOWN ETOOMANYREFS ETIMEDOUT ECONNREFUSED ELOOP ENAMETOOLONG EHOSTDOWN
-                    EHOSTUNREACH ENOTEMPTY EUSERS EDQUOT ESTALE EREMOTE ENOLCK ENOSYS EOVERFLOW
-                    ECANCELED EIDRM ENOMSG EILSEQ EBADMSG EMULTIHOP ENODATA ENOLINK ENOSR ENOSTR
-                    EPROTO ETIME EOPNOTSUPP ENOTRECOVERABLE EOWNERDEAD) %}
-    {% if LibC.has_constant?(value) %}
-      {{value.id}} = LibC::{{value.id}}
-    {% end %}
+  {% for value in LibC::Errno.constants %}
```

Comment:
This goes directly against https://github.com/crystal-lang/crystal/pull/8885#discussion_r390700078

Reply:
Why against? This PR is removing the duplicate definitions from this file. I responded to the link you mentioned here(Not realized it was a merged patch). Pasting the same comment here too. 

The existence of an Error number is one part of the problem. There is platform-specific value for some error codes. For example, `grep -R "EDEADLK" src/lib_c/*`

```
src/lib_c/aarch64-linux-gnu/c/errno.cr:        EDEADLK = 35
src/lib_c/aarch64-linux-musl/c/errno.cr:       EDEADLK = 35
src/lib_c/amd64-unknown-openbsd/c/errno.cr:    EDEADLK = 11
src/lib_c/arm-linux-gnueabihf/c/errno.cr:      EDEADLK = 35
src/lib_c/i386-linux-gnu/c/errno.cr:           EDEADLK = 35
src/lib_c/i386-linux-musl/c/errno.cr:          EDEADLK = 35
src/lib_c/i686-linux-gnu/c/errno.cr:           EDEADLK = 35
src/lib_c/i686-linux-musl/c/errno.cr:          EDEADLK = 35
src/lib_c/x86_64-darwin/c/errno.cr:            EDEADLK = 11
src/lib_c/x86_64-dragonfly/c/errno.cr:         EDEADLK = 11
src/lib_c/x86_64-freebsd/c/errno.cr:           EDEADLK = 11
src/lib_c/x86_64-linux-gnu/c/errno.cr:         EDEADLK = 35
src/lib_c/x86_64-linux-musl/c/errno.cr:        EDEADLK = 35
src/lib_c/x86_64-macosx-darwin/c/errno.cr:     EDEADLK = 11
src/lib_c/x86_64-openbsd/c/errno.cr:           EDEADLK = 11
src/lib_c/x86_64-portbld-freebsd/c/errno.cr:   EDEADLK = 11
src/lib_c/x86_64-unknown-freebsd/c/errno.cr:   EDEADLK = 11
src/lib_c/x86_64-windows-msvc/c/errno.cr:      EDEADLK = 36
```

Reply:
I think the best we can do is by merging all the common error codes into one place and define conflicting error codes in respective platform-specific files. I am not sure if I can pick that task now.

Reply:
Why against: previously the file had the full list of errnos, now it doesn't. The API is entirely derived from platform-specific details. There was at least a list of possible values even if some of them are missing.

Reply:
(i.e. the current state is not ideal, but this PR would take it farther away from ideal)

Reply:
> Why against: previously the file had the full list of errnos, now it doesn't. The API is entirely derived from platform-specific details. There was at least a list of possible values even if some of them are missing.

> previously the file had the full list of errnos,

File had full list of errnos, but not all included in `Enum` because of `has_constant` check.

> There was at least a list of possible values even if some of them are missing.

Now also it will be available, because the final Enum will have it specific to that platform.

Problems with earlier version:

- If a error number exists in platform-specific file and not present in `src/errno.cr` file then that will not be included in the final `Errno` enum(Because of the `has_constant?` check)
- Error numbers duplication - If new error number is added to platform-specific then an entry is required in `src/errno.cr` file too.

I don't think we can have a platform-independent list of Error codes.

Reply:
> Error numbers duplication

That's not a problem, btw. It adds no significant maintenance burden.

Reply:
But why? I think I am missing something. Please help me to understand that why do you think this patch is causing more problems?

Doc is generated based on the platform where the doc gen command is run, which shows all Enum members specific to that platform(https://crystal-lang.org/api/0.35.1/Errno.html). This is the case with all other programming languages too(For ex: Python is listing error numbers specific to Linux(https://docs.python.org/3/library/errno.html), but actual Python build will have platform specific error numbers, to check this: `python3 -c 'import errno; print(errno.ENODATA)'` in both Mac and Linux)

Reply:
https://github.com/crystal-lang/crystal/pull/8885#discussion_r390700078

> We should either move this enum into Crystal:: so it is not public or have it be always the same list of errno despite the platform: this enum must have the same members regardless of platform.

I don't think we can combine every errno to one file. As I mentioned earlier, we can merge all common errors to one place but not all.

Reply:
@oprypin please read the latest comment from @waj, we decided to do it the way it's done in this PR. Please let's not confuse the author.

Reply:
I think that comment is just misguided. Though indeed, I hadn't noticed it.

I would be curious who's "we", and would appreciate more consideration towards these prior opinions of people who have more context on cross-platform support.

Reply:
ah, sorry, let me get out of this discussion, ignore what I said

As always, I think it's bad that people send PRs before discussing things. @aravindavk could you open an issue to discuss this? That way you don't need to write code before reaching a conclusion

Reply:
@asterite Thanks. I opened an issue and updated the issue number in the description and commit message. Initially, I thought this a simple bugfix with a few more error codes. Based on more feedback, more refactoring added. This is my first contribution to the Crystal community, In the future, I will open the issue before sending PR. Thanks.


@oprypin 

> I would be curious who's "we", and would appreciate more consideration towards these prior opinions of people who have more context on cross-platform support.

I think he referred to the participants in this PR.

I didn't see any explanation about why this PR is a problem. The other issue you mentioned is about providing platform-independent Exceptions as much as possible for the File errors. More common errors are covered with that PR. That doesn't mean that there will be no need for checking granular error numbers.

With this PR I am trying to address the missing error codes with the existing infra. If you have a proposal to change the errno architecture itself. Please propose. I don't think this patch needs to solve all the issues related to the way of handling error numbers.

Please explain the issues introduced by this PR. Thanks.

Reply:
This PR initially just added error codes, which is probably fine by itself. https://github.com/crystal-lang/crystal/pull/9523#issuecomment-653021203 suggested to add a tangentially related refactoring to this PR. However, that suggestion would bring in non-trivial changes that have not been previously agreed upon -- in fact, there was some prior opposition to this in https://github.com/crystal-lang/crystal/pull/8885#discussion_r390700078 (not the entire PR, just the comment thread). Instead such a change should first be discussed and then (upon agreement) sent in a separate pull request.

Reply:
@oprypin let me clarify a few things.

- After receiving the review comments by @waj, I realized that adding new error codes to `src/errno.cr` is not enough. I had to add the same error numbers to respective platform-specific files as well. If I add only to `src/errno.cr` without adding to platform-specific files then it means same as not adding any new code(Because of `has_constant` check)
- You are mixing two things here. One is adding error codes to the `Errno` and the other one is hiding these errors and introducing common Exception classes(That is this comment https://github.com/crystal-lang/crystal/pull/8885#discussion_r390700078). 
    - This PR is fixing the first issue without changing the existing behavior
    - To make this enum Private as mentioned in the linked issue, all the required custom exceptions should be implemented. That will be a new feature altogether, which I am not planning to touch.
    - Even to achieve the same, we need these missing error numbers in the `Errno`(Private or Public)
- The number of lines change is due to sorted list of enums in respective platform-specific files.
- @waj's suggestion is only to add a namespace to error numbers just to separate the other constants exists in LibC

The changes in this PR is not changing any behavior. Let me know if I missed anything

Reply:
This PR deletes the list of constants from the file `src/errno.cr` (which we are currently commenting on), replacing it with "automation". I agree that technically that brings no change in existing behavior. But it moves us away from having an explicitly defined API, in addition to being a refactor that is unnecessary for just adding more error codes. So I think this code change should not be done, because then a proposal such as #9573 would need to re-add that list of constants anyway, so why go back and forth.

Reply:
For the record, this is how I would want the current PR to look (just adding constants):
https://github.com/oprypin/crystal/commit/467061ef03b277d3b6fc8b482c9bd4a59205ff47

Or, if merged after #9573:
https://github.com/oprypin/crystal/compare/0f06db70861f16965e54c5855a9d070134aab2bc...c5fc39dcb826d7c7df32f29a43ef91e705ea9dc6

Reply:
I liked the @asterite's suggestion to merge all error numbers to a single place(with `-1` for missing in specific platform).    That will further reduce duplications. Doc issues will still be their(showing the wrong value or -1 when doc generated in one platform).

I will wait for the response from @asterite before starting any further work on this.

Reply:
@aravindavk FYI initially there was one file with conditionals defining the errnos, but it was explicitly decided to split it up as it is currently.
See https://github.com/crystal-lang/crystal/pull/2442/files#diff-b67ae37b7956e573f1a5e19006b2615a (expand diff on errno.cr).

But I'm quite sure that that's not what [@asterite commented](https://github.com/crystal-lang/crystal/issues/9572#issuecomment-653767245). (I hope I'm not misunderstanding your understanding of what @asterite suggested......). I think that https://github.com/crystal-lang/crystal/issues/9572#issuecomment-653755214 and https://github.com/crystal-lang/crystal/issues/9572#issuecomment-653767245 are in line with #9573.

And again, this duplication is not seen as problematic; your goal of avoiding it shouldn't be a goal of itself.

I can try to explain why explicitly defining these constants in 2 places makes sense. One place describes existing APIs of a particular platform, and the other place explicitly describes Crystal's API, just that the API happens to be designed to match those platform APIs.